### DISCUSSION TITLE
Middleware doesn't run during tests

### DISCUSSION THREAD

**Discussion by bkilshaw** [livewire/livewire#4948] [7. Bugs (archived)]
I have a multi-tenant application that uses subdomains to identify the tenant. I have a route group that scopes the routes to the tenant.
```php
Route::domain('{tenant:subdomain}.'.config('app.base_url'))->group(function () {
   // routes
});
```
Because Livewires routes are registered outside this group I need a custom middleware for Livewire to configure some Tenant settings. 
```php
class ConfigureTenantForLivewire
{

    public function handle(Request $request, Closure $next)
    {
        $tenantSubdomain = Str::before(
            $request->getHost(),
            sprintf('.%s',config('app.base_url'))
        );

        $tenant = Tenant::where('subdomain', $tenantSubdomain)->firstOrFail();

        $request->route()->parameterNames[] = 'tenant';
        $request->route()->domain('{tenant:subdomain}.'.config('app.base_url'));
        $request->route()->setParameter('tenant', $tenant);

        $livewireDirectory = Config::get('livewire.temporary_file_upload.directory');
        Config::set('livewire.temporary_file_upload.directory', sprintf('%s/%s', $tenant->uuid, $livewireDirectory));

        return $next($request);
    }
}
```
Then in my config/livewire.php I have added this middleware to to the middleware_group:
```php
    'middleware_group' => [
        \App\Http\Middleware\ConfigureTenantForLivewire::class,
        'web'
    ],
```
This all runs great, except when I'm running tests.
```php
test('user can upload new logo', function () {
    Storage::fake('s3');
    Storage::fake('tmp-for-tests');

    $logo = UploadedFile::fake()->image('logo.png');

    $tenant = Tenant::first();
    expect($tenant->logo)->toBeNull();

    livewire(TenantSettings::class)
        ->set('logo', $logo)
        ->call('save')
        ->assertHasNoErrors();
});
```
What I've found in tests is that the configured middleware never run, so my tenant settings (and livewire-tmp path) never get configured as expected. This means that the test files don't go to the right places, and the tests fail.

Is there any reason why the configured middleware aren't being run during tests? Is this expected behaviour? Is there any way to make them run during tests?

Edit: Just to confirm; I get that the test itself isn't making any HTTP calls, but in reality the ->set() and ->call() methods both mimic actions that result in HTTP calls in a live application which get run through the configured middleware. By not having the middleware run on this actions in tests the test doesn't accurately reflect what happens when the code is executed by an actual person.

**Comment by joshhanley** [comment]
@bkilshaw hmm yeah the reason it doesn't is because it's not testing the route, it's only testing the component in isolation.

You will need to load a tenant manually (not sure if that will work with domain based tenants). But for example, I use a session variable to keep track of which tenant my user is logged in as.

So in my test, I generally set 

```php
$tenant = TenantFactory::create();
$user = UserFactory::forTenant($tenant)->create();

$this->actingAs($user)->withSession(['tenant_id' => $tenant->id]);

// Test stuff
```

Not sure if you can do something similar in your app.

Hope this helps!

**Reply by bkilshaw** [reply]
If that's the case then it feels like it's not really testing the component thoroughly. In reality there's no way to interact with a component without the interactions being run through the middleware.

```php
livewire(TenantSettings::class)
  ->set('logo', $logo)
  ->call('save')
  ->assertHasNoErrors();
```
What I really want to know when I run this is; will a user get any errors? If this doesn't simulate the calls (set/call) being routed through the middleware, which will always happen when a user interacts with the component, then it's not accurately testing the component.

TLDR; if `->set()` and `->call()` methods don't get routed through the middleware during tests then the test doesn't accurately reflect reality and leaves room for bugs.

Thoughts?

**Reply by renky** [reply]
Since this is unanswered for a long time now and I'm also struggeling with that, I just want to agree with @bkilshaw ...

There might also be reasons to test the component in total isolation but with that in mind I think Component-Only-Testing gets much less usefull.

In my situation I'm also in a multi-tenant-application and in my case I switch the tenant on login (additionally to the middleware detection) - so I didn't run into this issue, because "actingAs" implicitly fixed this issue...

Now I added a second Middleware where I try to detect the super-admin-user - this is done to reduce database-requests, because on several places it is asked "are you a superuser" - so I tried to add this to a middleware and came to the conclusion that it never runs in any of the component-tests... For sure - you can disucuss about my solution - there are several other possibilities to do that, for example with a session-variable, but thats not the point here... Since there is even a middleware-setting in livewire-config I'd really expect that this middlewares are also used in all testcases - even in Component-Isolation?! 

Am I wrong?

**Reply by joshhanley** [reply]
@renky best way to think of `Livewire::test()` is that they are unit tests, not integration tests. Hence you are unit-testing the component itself. As such, you need to prepare things like the user, and whether they are a super user or not, in the test before you call the component.

If you want a full end-to-end test, I'd recommend using Laravel Dusk to ensure everything works as expected as that also includes Livewire's javascript front end.

As it stands at the moment, there is no way to test an endpoint, like Laravel's `$this->get(...)...` that also interacts with the component class.

Hope this helps!

**Comment by PhillipMwaniki** [comment]
How are you currently able to set custom middleware for queries that run within a livewire component? I have a multitenant application that sets the connection at the application layer perfectly, but it doesn't work with livewire. Livewire 3 got rid of the middleware_group config.

I am using Livewire 3 with Laravel 10.
